openlibrary/plugins/openlibrary/js/carousel/Carousel.js

```diff
@@ -103,7 +103,7 @@ const Carousel = {
                 '"aria-hidden="false" role="option">' +
                 '<div class="book-cover">' +
                   '<a href="'}${work.key}" ${isClickable}>` +
-                    `<img class="bookcover" width="130" height="200" title="${
```

Comment:
We should keep these attributes.

More info on why here: https://www.smashingmagazine.com/2020/03/setting-height-width-images-important-again/

Reply:
These attributes break the solution / might be the problem in the first place. The book cover image aspect ratios vary. 

As far as performance is concerned, I appreciate the extra work that has to be done to respond to varying image sizes.
`object-fit: contain` solves the aspect-ratio issue, but you run into various UI problems around having an `img` tag's image almost being a child of the tag... box shadow, for a start, will size around the 130x200 tag instead of its contained (e.g.) 130x50 image.

As far as 'layout shifts' are concerned, all the space that the images will take are reserved/defined by their various parents (e.g. `div.book-cover` with height 200px). (There is one horizontal layout shift on the carousel, but it's a separate issue.)

Reply:
Thanks for explaining. 

Could we access the actual width and height values? Are they available here?

Reply:
What's the idea?

Reply:
The extra work performed by the browser, when without these set image dimensions, is calculating the aspect ratio & dimension which hits our constraint first.

contraints being 
```
img.bookcover {
    max-width: 100%;
    max-height: 200px;
}
```

For a known image, I can understand the optimisation by defining these HTML attributes for the browser to use. For a varying image, I don't know of the fix. I'd love to find out how a dev can solve this. By your linked article, maybe there are hacks, but no good soln.

Reply:
Yeh probably out of scope for this ticket.

@mekarpeles we have cover.type and cover.id variables  in [the search API](https://openlibrary.org/search.json?q=key%3A%28%2Fworks%2FOL450777W+OR+%2Fworks%2FOL362289W+OR+%2Fworks%2FOL4077051W+OR+%2Fworks%2FOL2715009W+OR+%2Fworks%2FOL2205289W+OR+%2Fworks%2FOL158953W+OR+%2Fworks%2FOL4662884W+OR+%2Fworks%2FOL222799W+OR+%2Fworks%2FOL5859708W+OR+%2Fworks%2FOL19659784W+OR+%2Fworks%2FOL2765935W+OR+%2Fworks%2FOL15834136W+OR+%2Fworks%2FOL513969W+OR+%2Fworks%2FOL98501W+OR+%2Fworks%2FOL464991W+OR+%2Fworks%2FOL8193418W+OR+%2Fworks%2FOL61324W+OR+%2Fworks%2FOL1870400W+OR+%2Fworks%2FOL50829W+OR+%2Fworks%2FOL66531W+OR+%2Fworks%2FOL5717098W+OR+%2Fworks%2FOL61921W+OR+%2Fworks%2FOL5475081W+OR+%2Fworks%2FOL875437W+OR+%2Fworks%2FOL6034514W+OR+%2Fworks%2FOL523452W+OR+%2Fworks%2FOL7711724W+OR+%2Fworks%2FOL1854080W+OR+%2Fworks%2FOL9347808W+OR+%2Fworks%2FOL2676023W+OR+%2Fworks%2FOL6218070W+OR+%2Fworks%2FOL10432709W+OR+%2Fworks%2FOL804244W+OR+%2Fworks%2FOL12497W+OR+%2Fworks%2FOL77792W+OR+%2Fworks%2FOL2721005W+OR+%2Fworks%2FOL4661335W+OR+%2Fworks%2FOL831059W+OR+%2Fworks%2FOL2731827W+OR+%2Fworks%2FOL21522W+OR+%2Fworks%2FOL482313W+OR+%2Fworks%2FOL97440W+OR+%2Fworks%2FOL3943151W+OR+%2Fworks%2FOL3521874W+OR+%2Fworks%2FOL2715015W+OR+%2Fworks%2FOL66544W+OR+%2Fworks%2FOL433123W+OR+%2Fworks%2FOL2068683W+OR+%2Fworks%2FOL6322288W+OR+%2Fworks%2FOL1971683W+OR+%2Fworks%2FOL1069667W+OR+%2Fworks%2FOL2438133W+OR+%2Fworks%2FOL17272376W+OR+%2Fworks%2FOL16482241W+OR+%2Fworks%2FOL15860364W+OR+%2Fworks%2FOL151996W+OR+%2Fworks%2FOL6740249W+OR+%2Fworks%2FOL15040422W+OR+%2Fworks%2FOL16069155W+OR+%2Fworks%2FOL508163W+OR+%2Fworks%2FOL3291229W+OR+%2Fworks%2FOL61003W+OR+%2Fworks%2FOL98491W+OR+%2Fworks%2FOL5888555W+OR+%2Fworks%2FOL5827913W+OR+%2Fworks%2FOL17933404W+OR+%2Fworks%2FOL1095427W+OR+%2Fworks%2FOL54915W+OR+%2Fworks%2FOL13114894W+OR+%2Fworks%2FOL24338132W+OR+%2Fworks%2FOL1872916W+OR+%2Fworks%2FOL15840480W+OR+%2Fworks%2FOL184431W+OR+%2Fworks%2FOL2940316W+OR+%2Fworks%2FOL2647505W+OR+%2Fworks%2FOL259028W+OR+%2Fworks%2FOL14915863W+OR+%2Fworks%2FOL29462W+OR+%2Fworks%2FOL1734184W+OR+%2Fworks%2FOL675449W+OR+%2Fworks%2FOL18591W+OR+%2Fworks%2FOL221675W+OR+%2Fworks%2FOL5704260W+OR+%2Fworks%2FOL15717066W+OR+%2Fworks%2FOL3863998W+OR+%2Fworks%2FOL2619717W+OR+%2Fworks%2FOL64151W+OR+%2Fworks%2FOL12826W+OR+%2Fworks%2FOL547889W+OR+%2Fworks%2FOL66534W+OR+%2Fworks%2FOL15952404W+OR+%2Fworks%2FOL2155632W+OR+%2Fworks%2FOL69503W+OR+%2Fworks%2FOL61215W+OR+%2Fworks%2FOL112890W+OR+%2Fworks%2FOL66562W+OR+%2Fworks%2FOL15837476W+OR+%2Fworks%2FOL15178362W+OR+%2Fworks%2FOL2046569W+OR+%2Fworks%2FOL2031517W%29&has_fulltext=true&wt=json&limit=20&offset=20). How hard would it be to also surface cover width and height for this use case?

Reply:
Not sure if this is exactly what you need, but an image's height and width can be fetched from the covers API:

**Using Cover ID:**
http://covers.openlibrary.org/b/id/8747029.json

**Using OLID (2 ways):**
http://covers.openlibrary.org/b/olid/OL17791670M.json
http://covers.openlibrary.org/b/query?olid=OL17791670M&details=true

Unfortunately, height and width fields are not guaranteed to have non-null values:
http://covers.openlibrary.org/b/id/472861.json
http://covers.openlibrary.org/b/id/472861-M.jpg

Reply:
So yeh.. that was what I was looking for, but we'd also need an internal python version for the template below since this needs to be outputted by the template.